### ISSUE TITLE
Deprecate PodNodeSelector admission plugin

### ISSUE THREAD

**Issue by ahg-g** [kubernetes/kubernetes#113474]
### What would you like to be added?

Deprecate PodNodeSelector admission plugin and the associated annotation `scheduler.alpha.kubernetes.io/node-selector`

https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/podnodeselector/admission.go

### Why is this needed?

This plugin has been in alpha for a long time and never graduated. There are other ways now to support a similar behavior using mutating webhooks.

/sig scheduling

**Comment by sftim** [comment]
If / when we deprecate this, it'd be nice to document what people who rely on it can do to migrate. Yes, it's alpha but _somewhere_ is somebody who depends on the feature.

**Comment by kerthcet** [comment]
> If / when we deprecate this, it'd be nice to document what people who rely on it can do to migrate. Yes, it's alpha but _somewhere_ is somebody who depends on the feature.

We'll have a release note about this change. Where to place the document? The well-known labels page but with a deprecation warning?

**Comment by kerthcet** [comment]
We're approaching code freeze now, for the sake of the stable goal, I'll follow this until v1.27.

**Comment by sanposhiho** [comment]
> Where to place the document? The well-known labels page but with a deprecation warning?

Currently, ↓ has the description about this label. We should add a deprecation notice here at least.
https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#configuration-annotation-format

**Comment by sanposhiho** [comment]
Theoretically, we could remove this feature right now because it's alpha.
But, given that it's a quite old feature (from [2016](https://github.com/kubernetes/kubernetes/pull/24980)?) and many users may already use it, it'd be kind if we could just announce the deprecation for now and wait for a few releases to remove it completely.

**Comment by ahg-g** [comment]
> We're approaching code freeze now, for the sake of the stable goal, I'll follow this until v1.27.

sg, thanks.

**Comment by kerthcet** [comment]
Revisit this issue, actually, we're depending on this with `PodTolerationRestriction` to force pods in the same namespaces running in the specified nodes, so I'm considering to graduate this to beta instead? Along with `PodTolerationRestriction`. WDYT? @ahg-g

**Comment by ahg-g** [comment]
We should deprecate both plugins, `PodTolerationRestriction` seems to be in alpha too: https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#podtolerationrestriction

**Comment by kerthcet** [comment]
Yes, they're all in alpha level, but that's may because no one is following these two features as it behaves well, but doesn't mean no end-users. 

Our product relies on these two plugins to force pods with-in the same namespace running on the specified nodes, it's like a hard constraint. And this feature is required by several of our customers, so from the users POV, we hope to keep them as they are.  

Or do you think we should get some feedbacks from the community, email survey etc.?

**Comment by sftim** [comment]
> Yes, they're all in alpha level, but that's may because no one is following these two features as it behaves well, but doesn't mean no end-users.

If we _deprecate_ something, it's not removed immediately. Deprecation means “you should stop using this feature”. When we assert that, you should stop. The deprecation lets us make sure that Kubernetes overall is well maintained, and that's a quality that our end users value.

People who wanted to make an alternative to those deprecated admission controllers could use other Kubernetes extension points to do so. There are other ways now to support a similar behavior using mutating webhooks (stable), or perhaps via [scheduling plugins](https://kubernetes.io/docs/concepts/scheduling-eviction/scheduling-framework/). 
The deprecation period allows for that migration to happen.

(If you're _relying_ on an alpha feature of Kubernetes, that's absolutely your prerogative, but as a project I'm confident that we've made the story clear. Alpha level makes a strong and widely recognised assertion about the stability of that feature and the guarantees we're **not** making about it remaining part of Kubernetes).

In the ecosystem, you can find [certified service providers](https://landscape.cncf.io/card-mode?category=kubernetes-certified-service-provider&grouping=category) for Kubernetes who can help you adapt.

---

Alternatively, you _could_ propose a KEP to move these admission mechanisms to stable. However, I think the community would recommend an out-of-tree implementation using admission webhooks, even if the annotation for that becomes an official Kubernetes one.

**Comment by kerthcet** [comment]
> If we deprecate something, it's not removed immediately. Deprecation means “you should stop using this feature”. When we assert that, you should stop. The deprecation lets us make sure that Kubernetes overall is well maintained, and that's a quality that our end users value.

For alpha level features, we can remove them directly https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecation

But considering this is a feature came to reality a long time ago, adding a warning before removal may not surprise end-users a lot. 

> There are other ways now to support a similar behavior using mutating webhooks (stable), or perhaps via [scheduling plugins](https://kubernetes.io/docs/concepts/scheduling-eviction/scheduling-framework/).

If we have one approach in kubernetes that meets users' requirements, why should we remove them instead asking the users to implement themselves, which we say reimplementing the wheels. And I don't think scheduler-plugins can achieve this, it's more about scheduling.

> (If you're relying on an alpha feature of Kubernetes, that's absolutely your prerogative, but as a project I'm confident that we've made the story clear. Alpha level makes a strong and widely recognised assertion about the stability of that feature and the guarantees we're not making about it remaining part of Kubernetes).

Well, it likes a chicken-egg problem, every feature incubates from alpha level, we use this feature and we also want to enhance its capacity like toleration for DS pods. But agree with you that alpha-level features not always stable.

After all, if this feature is widely used(This is not sure yet) by the kubernetes community, but because it's alpha level(not every users are developers), and we remove  them directly,  I have no idea whether this is the right direction for evolution.

Thanks for your detailed comments @sftim

**Comment by lory-to** [comment]
> Yes, they're all in alpha level, but that's may because no one is following these two features as it behaves well, but doesn't mean no end-users.

I'm among the community using it and they provide an elegant way to manage our workloads.
So yes, please ask community I'm sure we're a lot taking advantage of them!

**Comment by CharmanderJieniJieni** [comment]
If I do a search on open/closed issues with keyword `PodNodeSelector` in this repository, people are suggested to use PodNodeSelector as a namespace scoped node selector. And then I see this threat intending to remove this feature. 

I am a bit confused what would be the suggested approach now. Namespace level node isolation is a legit use case for K8s administrator, when you don't have much control on a pod template inside the namespace while you still want to enforce pods not being scheduled in unintended nodes. Even with taint/toleration applied an end users can still bypass the security boundary to schedule the pod in a privileged node by adding toleration if they know the target taint. 

I did a some research and seems the only alternative which is not in alpha is to use Dynamic Admission Control and write our own piece of Admission Webhook. Since a lot of people are relying on this feature, why not moving it to GA?

**Comment by sftim** [comment]
> Since a lot of people are relying on this feature, why not moving it to GA?

_This plugin has been in alpha for a long time and never graduated._ That's a sign that we don't have volunteers interested in moving the work forward.

As _there are other ways now to support a similar behavior using mutating webhooks_ I think the deprecation is the right thing to do, for the existing in-tree code.

---

For anyone who currently relies on `scheduler.alpha.kubernetes.io/node-selector` and would like to see a generally available and supported `scheduler.kubernetes.io/node-selector` (or similar) annotation - I believe that this **can** happen. At least theoretically. Even with the in-tree admission controller deprecated and eventually removed, Kubernetes can instead provide out of tree [[_jargon explainer_](https://www.wordsense.eu/out-of-tree/)] support for using it.

You can either write an implementation yourself and donate the code to Kubernetes, or you can sponsor existing org members to write an out of tree implementation. For example, an admission webhook helper, or a controller that manages [ValidatingAdmissionPolicyBindings](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/#what-is-validating-admission-policy) based on the annotations set for a namespace.

Personally, I don't think it's worth the effort. Anyone relying on the alpha feature already knows the [limited guarantees about stability](https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecation) (ie: the next minor release could remove it), or shouldn't be surprised by them. As such, folks'll have to migrate to something else anyway, and it may as well be to something that's already past alpha.

**Comment by sftim** [comment]
BTW, we will write the documentation for deprecation. The deprecation hasn't happened yet. When it does, we'll accompany that with at least some documentation. Help from other volunteers would of course be very welcome.

**Comment by kerthcet** [comment]
Still prefer to graduate this feature to beta, I'll email the sig scheduling folks for feedbacks and maybe bring this to the next sig scheduling meeting.

**Comment by alculquicondor** [comment]
Outcome from SIG meeting:

What is clear is that we cannot keep maintaining the `scheduler.alpha.kubernetes.io/node-selector` annotation as it has `alpha` in the name. However, since the feature has been there for so long, we can adhere to the beta deprecation policy: remove 3 releases after marked as deprecated https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-parts-of-the-api

The next question is what is the correct API to follow.

What follows are my thoughts after the SIG meeting.

I can see the following options:
1. Keep as annotation `scheduler.kubernetes.io/node-selector`. Probably the least desirable option as annotations are arbitrary strings.
2. Remove the functionality altogether. A webhook is simple enough or users could use policy agents such as gatekeeper.
3. Implement the functionality as an API field. `LimitRange` is the closest API I can think of. Adding a new resource definition might be overkill, but maybe the functionality can be added to the Namespace object. However this could cause a lot of more feature requests about defaulting of other Pod fields per namespace, to which the answer is probably to use a policy agent.

@kerthcet and others that are using the feature: can you share why you think this feature is important on its own (why not do the same for other scheduling fields) and why a policy agent is not a viable solution. What is your use case?

**Comment by alculquicondor** [comment]
cc @liggitt as API reviewer to weigh in

**Comment by kerthcet** [comment]
We can achieve `exclusive nodes per namesapce` with `podNodeSelector` and `PodTolerationRestriction`, we want pods belong to a namespace to run on specifically labeled and tainted nodes, pods from other namespaces can not be located to these nodes. 

> Probably the least desirable option as annotations are arbitrary strings.

I agree annotations is a good choice for experiments but not a long-term plan.

> A webhook is simple enough or users could use policy agents such as gatekeeper

We can use gatekeeper as mutation by injecting tolerations and nodeSelectors, but I don't know whether Rego supports the conflict validation as we do today in admission plugins. Even it supports, not every cluster installed the gatekeeper, maintaining a new component always means a new debt, and `namespace level node isolation` is a general scheduling requirement the community asks for, kubernetes should be the best place to support this, gatekeeper in my mind is better for customized demands for each company? I guess so...

Maybe a new API is not a bad idea like `LimitRange` and `ResourceQuota`, used for defaulting fields, also like schedulerName?

**Comment by liggitt** [comment]
> cc @liggitt as API reviewer to weigh in

specifically for option 3 ("maybe the functionality can be added to the Namespace object"), we wouldn't add pod-specific fields to the namespace API.

**Comment by alculquicondor** [comment]
Still, I'm having a hard time justifying this as an in-tree feature.

**Comment by alculquicondor** [comment]
We could wait for [Mutating admission with CEL](https://github.com/kubernetes/enhancements/issues/3775) and remove the plugin.

**Comment by liggitt** [comment]
> We could wait for [Mutating admission with CEL](https://github.com/kubernetes/enhancements/issues/3775) and remove the plugin.

yeah, that would be a useful scenario to run the mutating CEL proposal through... merging node selectors could be ... tricky

**Comment by ahg-g** [comment]
I am still in the camp that this feature doesn't need to be upstream and can be implemented as a webhook; the question is if an API that has been in alpha for such a long time, do we consider it defacto supported and should be graduated?

**Comment by kerthcet** [comment]
> We could wait for https://github.com/kubernetes/enhancements/issues/3775 and remove the plugin.

CEL mutation can somehow achieve similar behaviors like when finding a pod in a specific namespace, we'll inject node labels and tolerations, however, we may not validate the conflicts here.

also cc @knelasevero who has similar use cases. Will this meet your needs already?

**Comment by alculquicondor** [comment]
>  however, we may not able to validate the conflicts here.

Can you elaborate?

**Comment by knelasevero** [comment]
> also cc @knelasevero who has similar use cases. Will this meet your needs already?

So, this is enabled by default in openshift. So use cases for this are from our customers, not ours directly. We can't be totally sure of every way they are using it.

But yeah, we would have customers in similar situations as you describe.

> we'll inject node labels and tolerations, however, we may not able to validate the conflicts here.

What do you mean by conflicts?

@ingvagabund, @soltysh can you share more context here? You've been around longer.

**Comment by knelasevero** [comment]
> the question is if an API that has been in alpha for such a long time, do we consider it defacto supported and should be graduated?

I'd say yes. This have been around for a long time with no big complaints and with almost no changes, and (I think) enough people have been using it often enough to justify graduation. A bit of speculation here, but I think it's safe to assume.

**Comment by alculquicondor** [comment]
> This have been around for a long time with no big complaints and with almost no changes

I don't think this is enough justification to graduate.

1. The current annotation has `alpha` in the name. As we graduate we have to change the API. I don't think just removing the `alpha` substring is enough to pass API reviews, as we have higher standards now.
2. This admission controller was added at a time when webhooks or policy agents where not mature. The ecosystem is wildly different now.

Because users will have to take action to continue to "have" this feature, I think we should maximize API best practices.

Additionally, having the ability to control node affinity per namespace opens up the possibility for people to ask: "can you also add X per namespace"? And that's a big NO for me.

I much rather:
1. Declare this deprecated for 3 or 4 releases.
2. Wait for CEL mutation to be available.
3. Remove this plugin.

**Comment by knelasevero** [comment]
That's fair.

If changes would be needed anyways, causing disruptions to adhere to the new better standards, better to go all in and deprecate then.

> yeah, that would be a useful scenario to run the mutating CEL proposal through... merging node selectors could be ... tricky

Still curious about how tricky this will be. CEL kinda needs to be expressive enough to accommodate this use-case. I guess we are still a bit far from knowing how easy it would be to fit this in.

**Comment by alculquicondor** [comment]
@andrewsykim do you have any early design about the kind of possibilities CEL mutation could offer?

To my untrained eye, adding a nodeSelector based on the namespace sounds rather like a basic task for CEL.

**Comment by knelasevero** [comment]
@kerthcet are the conflicts you mentioned the same as what @liggitt phrased as `merging node selectors could be ... tricky`?

**Comment by alculquicondor** [comment]
@andrewsykim pointed me to the draft KEP by @jpbetz :) https://github.com/kubernetes/enhancements/pull/3963

**Comment by jpbetz** [comment]
> Andrew Sy Kim do you have any early design about the kind of possibilities CEL mutation could offer?
> 
> To my untrained eye, adding a nodeSelector based on the namespace sounds rather like a basic task for CEL.

The mutation looks like it could be trivially handled by https://github.com/kubernetes/enhancements/pull/3963

We are adding namespace object access to ValidatingAdmissionPolicy in 1.28, so mutating will have access to the namespace as well when we introduce it, hopefully in 1.29.

**Comment by ahg-g** [comment]
Even without CEL, webhooks are a very reasonable path forward; can't Openshift implement the exact same semantics you have now with the same "api" using a webhook without the customers even noticing?

**Comment by kerthcet** [comment]
> @kerthcet are the conflicts you mentioned the same as what @liggitt phrased as merging node selectors could be ... tricky?

Sorry, just back from holiday. Yes, they're the current logics these admission plugins have. E.g. pod has a label like `foo:bar` and we want to inject a new label named like `foo:fuz`, with the admission plugins, pod creation will be rejected, I have no idea whether the CEL can achieve, rather than just injecting a label.

**Comment by kerthcet** [comment]
> Even without CEL, webhooks are a very reasonable path forward; can't Openshift implement the exact same semantics you have now with the same "api" using a webhook without the customers even noticing?

On my poor opinion, both we can do this easily like copy-paste the implementation, but seems a certain of people need this, if the upstream can provide a solution, we can prevent reimplementing the wheel, webhook still needs some developing works.

**Comment by alculquicondor** [comment]
> I have no idea whether the CEL can achieve, rather than just injecting a label.

The feature is not fully designed yet. Can you bring this use case to the KEP? https://github.com/kubernetes/enhancements/pull/3963


>  if the upstream can provide a solution, we can prevent reimplementing the wheel

My question still is: when do we stop? If we accept this feature upstream, how can we tell future developers that we won't accept more policies based on Namespaces for other pod scheduling directives. What makes node selectors special, in comparison to tolerations, spreading, etc.

CEL could provide all the flexibility that users might need without being a maintenance burden for upstream.

**Comment by kerthcet** [comment]
I'm on the side of CEL, what I mean is if we do not have better alternatives we should not deprecate them directly.

> I much rather:
Declare this deprecated for 3 or 4 releases.
Wait for CEL mutation to be available.
Remove this plugin.

I'm glad with the path, so I'd like to mark it as deprecated in this release.

**Comment by jpbetz** [comment]
> It would be useful to confirm what is the expected timeline for CEL. Joe Betz, are you targeting alpha in 1.28?

Validating CEL admission: Alpha 1.27, Beta 1.28, GA: TBD
Mutating CEL admission: Alpha 1.29, Beta: TBD, GA: TBD

**Comment by ksyblast** [comment]
Hello everyone!
I am a kubernetes user and I am looking for a feature to dedicate specific nodes to a namespace. As I understand, I cannot rely anymore on PodNodeSelector as it's going to be deprecated soon. Could you please point me to the right direction how to implement this at the moment? Or is it quite safe to go now with PodNodeSelector and migrate to CEL Mutations further? I'm concerned that it may be not so easy to migrate then. Probably it makes sense to wait for the new feature first?
Thanks!

**Comment by alculquicondor** [comment]
We won't remove the alpha PodNodeSelector until CEL mutations are ready.

You can also use a webhook.

**Comment by janakirammeka** [comment]
We also same requirement to force  namespaced pods on specific nodes group.

**Comment by kerthcet** [comment]
/remove-lifecycle stale

CEL mutations is still un-beta now.

**Comment by alculquicondor** [comment]
It's actually targeting alpha for 1.31 now https://github.com/kubernetes/enhancements/issues/3962#issuecomment-1970296147

**Comment by alculquicondor** [comment]
cc @dom4ha
